### PR TITLE
Matrix connector

### LINKED ISSUE #662: Move Matrix connector into core
This issue covers adding the [Matrix connector](https://github.com/opsdroid/connector-matrix) to core.

## Background
A while ago we began moving connectors from external plugins into the core of the project (see #185 for more context). We started with [slack](https://github.com/opsdroid/opsdroid/blob/master/opsdroid/connector/slack/__init__.py) and [websockets](https://github.com/opsdroid/opsdroid/blob/master/opsdroid/connector/websocket/__init__.py) but need to go through all the other existing plugins and move them into the core.

## Steps

 - Make a new submodule directory in [`opsdroid.connector`](https://github.com/opsdroid/opsdroid/tree/master/opsdroid/connector) and copy the connector code over.
 - Update the [`requirements.txt`](https://github.com/opsdroid/opsdroid/blob/master/requirements.txt) with any dependencies from the connector if necessary.
 - Write tests for the connector. (See the [Slack connector tests](https://github.com/jacobtomlinson/opsdroid/blob/master/tests/test_connector_slack.py) for inspiration).
 - Copy the relevant information from the connector `README.md` into a [new documentation page](https://github.com/opsdroid/opsdroid/tree/master/docs/connectors).
 - Add the new page to the [mkdocs.yml](https://github.com/opsdroid/opsdroid/blob/master/mkdocs.yml).
 - Add to the [list of connectors](https://github.com/opsdroid/opsdroid/blob/master/docs/configuration-reference.md#connector-modules).
 - Add a deprecation notice to the old connector. (See [the slack connector](https://github.com/opsdroid/connector-slack))

### PR DESCRIPTION
**Pull request by SolarDrew**
# Port Matrix connector into core

Start of efforts to move the Matrix connector in the core library

Fixes #662 

- [x] Make a new submodule directory in opsdroid.connector and copy the connector code over.
- [x] Update the requirements.txt with any dependencies from the connector if necessary.
- [x] Write tests for the connector.
- [x] Copy the relevant information from the connector README.md into a new documentation page.
- [x] Add the new page to the mkdocs.yml.
- [x] Add to the list of connectors.
- [x] Add a deprecation notice to the old connector. (Pending in https://github.com/opsdroid/connector-matrix/pull/42)

## Status
**READY**

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# How Has This Been Tested?

- Existing tests pass
- New tests have been added and also pass

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by SolarDrew** [comment]
Note that I've ported the matrix connector across complete with its commit history, which I'm happy to purge if we don't want it cluttering up the place.

**Comment by jacobtomlinson** [comment]
Thanks for this! I see you've ticked the documentation related tasks but I can't see them in the diff.

**Comment by SolarDrew** [comment]
Oh, sorry, that probably means I'm a moron and forgot to commit them. I'll try and fix them and untick those tasks in the meantime.

File: requirements.txt [hunk 1]
```diff
@@ -12,3 +12,5 @@ websockets==7.0
 appdirs==1.4.3
 multidict==4.4.2
 motor==2.0.0
+git+https://github.com/tinloaf/matrix-python-sdk@groups
```
Comment (SolarDrew) [inline requirements.txt]: @Cadair is there a more stable position for this now that I can point to? Alternatively, @jacobtomlinson is there a way of specifying a repo as a requirement that doesn't break the tests?

Comment (Cadair) [reply requirements.txt]: This should be my async api package now? We don't need to rely on the groups branch for this.

Comment (jacobtomlinson) [reply requirements.txt]: Sadly we use the `requirements.txt` file in the `setup.py` too for dependencies, so a git style requirement is not supported. I recommend using (or publishing your own) package on pypi/conda-forge.

Comment (Cadair) [reply requirements.txt]: this line should just be `matrix-api-async`

**Comment by jacobtomlinson** [comment]
@Cadair @SolarDrew any chance you intend to pick this up again?

**Comment by SolarDrew** [comment]
Well I definitely _intend_ to... I'll try and do some more work on it this week.

**Comment by SolarDrew** [comment]
I force pushed to remove all the connector git history because I'm awful and you were going to squash it anyway. Tests still ongoing but all the code should be mostly in place now.

**Comment by FabioRosado** [comment]
Not stale, bad bot 😄

**Comment by jacobtomlinson** [comment]
Happy new year everyone! Any chance you can find some time to work on this?

**Comment by SolarDrew** [comment]
> Happy new year everyone! Any chance you can find some time to work on this?

Happy new year :) Yes, apologies for the slow progress on this. @Cadair is helping me out with some of the testing stuff, which was where I'd got stuck, so there should start to be some progress on this pretty soon, hopefully.

**Comment by FabioRosado** [comment]
🎉 This is awesome! Thanks for working on this I am merging this now 👍

**Pull request merged**